contracts/src/token/erc1155/extensions/metadata_uri.rs

```diff
@@ -0,0 +1,101 @@
+//! Optional URI Metadata of the ERC-1155 standard, as defined
+//! in the [ERC].
+//!
+//! [ERC]: https://eips.ethereum.org/EIPS/eip-1155#metadata-extensions
+
+use alloc::string::String;
+
+use alloy_primitives::{FixedBytes, U256};
+use alloy_sol_macro::sol;
+use openzeppelin_stylus_proc::interface_id;
+use stylus_sdk::stylus_proc::{public, sol_storage};
+
+use crate::utils::introspection::erc165::{Erc165, IErc165};
+
+sol! {
+    /// Emitted when the URI for token type `id` changes to `value`, if it is
+    /// a non-programmatic URI.
+    ///
+    /// If a [`URI`] event was emitted for `id`, the standard [guarantees] that
+    /// `value` will equal the value returned by [`IErc1155MetadataUri::uri`].
+    ///
+    /// [guarantees]: https://eips.ethereum.org/EIPS/eip-1155#metadata-extensions
+    #[allow(missing_docs)]
+    event URI(string value, uint256 indexed id);
+}
+
+sol_storage! {
+    /// URI Metadata of an [`crate::token::erc1155::Erc1155`] token.
+    pub struct Erc1155MetadataUri {
+        /// Used as the URI for all token types by relying on ID substitution,
+        /// e.g. https://token-cdn-domain/{id}.json.
+        string _uri;
+    }
+}
+
+/// Interface for the optional metadata functions from the ERC-1155 standard.
+#[interface_id]
+pub trait IErc1155MetadataUri {
+    /// Returns the URI for token type `id`.
+    ///
+    /// If the `id` substring is present in the URI, it must be replaced by
+    /// clients with the actual token type ID.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the contract's state.
+    /// * `id` - Token id.
+    fn uri(&self, id: U256) -> String;
+}
+
+// FIXME: Apply multi-level inheritance to export Metadata's functions.
+// With the current version of SDK it is not possible.
+// See https://github.com/OffchainLabs/stylus-sdk-rs/pull/120
+#[public]
+impl IErc1155MetadataUri for Erc1155MetadataUri {
+    /// This implementation returns the same URI for all token types.
+    /// Clients calling this function must replace the `id` substring with
+    /// the actual token type ID.
+    fn uri(&self, _id: U256) -> String {
```

Comment:
As I see to override this behavior consumer should basically rebuild `Erc1155MetadataUri` contract.
I'm curious @bidzyyys , could we consider to place this extension within `Erc1155` contract? Since in solidity it is implemented by default for `ERC1155` abstract contract...

Reply:
You are right @qalisander, but on the other hand we have the same case in Erc20 & Metadata extension - in our implementation it is separately. I would vote for leaving as it is right now. Wdyt? @ggonzalez94 @0xNeshi

Reply:
Sure. We just need to keep in mind that it will be a subject to refactor later

Reply:
A lot of things will have to be refactored once proper multi-inheritance is enabled, multi-`#[public]`, etc. This specific change will be a drop in the bucket. 
I vote we leave it